PurchasesCoreSwift/Purchasing/OfferingsManager.swift

```diff
@@ -0,0 +1,161 @@
+        }
+    }
+
+}
+
+private extension OfferingsManager {
+
+    func handleOfferingsBackendResult(with data: [String: Any], completion: ReceiveOfferingsBlock?) {
+        let productIdentifiers = extractProductIdentifiers(fromOfferingsData: data)
+
+        productsManager.products(withIdentifiers: productIdentifiers) { products in
+            let productsByID = products.reduce(into: [:]) { result, product in
+                result[product.productIdentifier] = product
+            }
+
+            if let createdOfferings = self.offeringsFactory.createOfferings(withProducts: productsByID, data: data) {
+                self.logProductsMissingIfAppropriate(products: productsByID, offeringsData: data)
+
+                self.deviceCache.cache(offerings: createdOfferings)
+                self.dispatchCompletionOnMainThreadIfPossible(completion,
+                                                              offerings: createdOfferings,
+                                                              error: nil)
+            } else {
+                self.handleOfferingsUpdateError(ErrorUtils.unexpectedBackendResponseError(), completion: completion)
+            }
+        }
+    }
+
+    func handleOfferingsUpdateError(_ error: Error, completion: ReceiveOfferingsBlock?) {
+        Logger.appleError(String(format: Strings.offering.fetching_offerings_error, error as CVarArg))
+        deviceCache.clearOfferingsCacheTimestamp()
+        dispatchCompletionOnMainThreadIfPossible(completion,
+                                                 offerings: nil,
+                                                 error: error)
+    }
+
+    func extractProductIdentifiers(fromOfferingsData offeringsData: [String: Any]) -> Set<String> {
+        guard let offerings = offeringsData["offerings"] as? [[String: Any]] else {
+            return []
+        }
+
+        let productIdenfitiersArray = offerings
+            .compactMap { $0["packages"] as? [[String: Any]] }
+            .flatMap { $0 }
+            .compactMap { $0["platform_product_identifier"] as? String }
+
+        return Set(productIdenfitiersArray)
+    }
+
+    func logProductsMissingIfAppropriate(products: [String: SKProduct], offeringsData: [String: Any]) {
+        guard !products.isEmpty,
+              !offeringsData.isEmpty else {
+            return
+        }
+
+        let productIdentifiers = extractProductIdentifiers(fromOfferingsData: offeringsData)
+        let missingProducts = Set(products.keys).intersection(productIdentifiers)
+
+        if !missingProducts.isEmpty {
+            Logger.appleWarning(Strings.offering.cannot_find_product_configuration_error)
+        }
+    }
+
+    func dispatchCompletionOnMainThreadIfPossible(_ completion: ReceiveOfferingsBlock?, offerings: Offerings?, error: Error?) {
+        if let completion = completion {
+            operationDispatcher.dispatchOnMainThread {
+                completion(offerings, error)
+            }
+        }
+    }
```

Comment:
this is good, we might wanna make it a part of `OperationDispatcher` so we can reuse it

Reply:
Firstly I thought the same, but I didn't do it because I'm not sure if this code should be managed by `OperationDispatcher`. I mean, being purists 😬 the `OperationDispatcher`'s responsibility is to dispatch blocks of code, I'm not sure if it should check if an object that is used inside the block is `nil`, do you know what I mean? Although if we are pragmatics, it'll allow us to avoid many `if let completion = completion {` 😓

But the main problem is the `completion` is not always the same, so we have to use generics and ObjC doesn't support them yet 😑 or use 'Any' and...I don't consider it as a real alternative.

WDYT?

Reply:
oh, right, forgot about that. I actually went through the same process when I wrote the OperationDispatcher in the first place 🤦